make/_shared/oci-build/image_tool/append_layers.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2023 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/google/go-containerregistry/pkg/v1/match"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/spf13/cobra"
+)
+
+var CommandAppendLayers = cobra.Command{
+	Use:   "append-layers oci-path [path-to-tarball...]",
+	Short: "Appends a tarball or directory to every image in an OCI index.",
+	Args:  cobra.MinimumNArgs(1),
```

Comment:
suggestion (non-blocking): Not something that should change for this PR although it's an important thing: I think changing files in-place like this is a bit scary with make (and generally) and it might be safer to write a temporary file first and then do the modification to that temporary file and save it with a new name.

The current flow means that a tool outside of make can't ever safely read/use the tarball on disk as-is because it can't know if it's in the process of being changed or not. It either has to check that make isn't running (and ensure it doesn't start running before reading the file) or else check the contents of the file before using it to ensure it was correctly updated.

This would imply ko building a temporary tarball (e.g. `$(oci_layout_path_x).beforeappend`) and then the append-layers script writing to `$(oci_layout_path_x)`

Reply:
The tool actually changes the `oci-path` which is a folder structure representing a multi-arch OCI image.
The `convert_to_docker_tar.go` file actually takes the `oci-path` folder as input and generates a tarball for the local architecture.
I agree that there is some work-to-be-done to simplify this area. There is also some hope in me still that some of these workarounds will be deprecated in the future in favor of upstream improvements.

Reply:
Ah fair enough! Thanks for clarifying